CA_DataUploaderLib/MCUBoard.cs

```diff
@@ -373,28 +389,43 @@ private bool TryReadOptionalCalibration(ref ReadOnlySequence<byte> buffer, out s
+            // if a device never sends a line for it to read.
+            int millisecondsTimeout = 3000;
+            using var cts = new CancellationTokenSource(millisecondsTimeout);
+            var token = cts.Token;
             foreach (var detector in customProtocolDetectors)
             {
-                while (watch.ElapsedMilliseconds < 3000) //only allow up to 3 seconds to detect a third party device
+                while (!token.IsCancellationRequested) //only allow up to 3 seconds to detect a third party device
                 {
-                    var result = await pipeReader.ReadAsync();
-                    var buffer = result.Buffer;
-
-                    var (finishedDetection, detectedInfo) = detector(buffer, portName);
-                    if (finishedDetection && detectedInfo != default)
-                        return detectedInfo; //confirmed it is a third party, leave pipeReader data not examined so next read processes data available right away
-                    if (finishedDetection && detectedInfo == default)
-                        break; //move to next third party. Note data is not set as examined in the pipeReader so ReadAsync returns inmediately with already read data.
-
-                    // set all data as examined as we need more data to finish detection
-                    pipeReader.AdvanceTo(buffer.Start, buffer.End);
-
-                    if (result.IsCompleted)
-                        return default; //should only happen if the connection was closed
+                    try 
+                    { 
+                        var result = await pipeReader.ReadAsync(token); 
+                    
+                        var buffer = result.Buffer;
+
+                        var (finishedDetection, detectedInfo) = detector(buffer, portName);
+                        if (finishedDetection && detectedInfo != default)
+                            return detectedInfo; //confirmed it is a third party, leave pipeReader data not examined so next read processes data available right away
+                        if (finishedDetection && detectedInfo == default)
+                            break; //move to next third party. Note data is not set as examined in the pipeReader so ReadAsync returns inmediately with already read data.
+
+                        // set all data as examined as we need more data to finish detection
+                        pipeReader.AdvanceTo(buffer.Start, buffer.End);
+
+                        if (result.IsCompleted)
+                            return default; //should only happen if the connection was closed
+                    } 
+                    catch (OperationCanceledException ex)
+                    {
+                        if (ex.CancellationToken == token)
+                        {
+                            CALog.LogColor(LogID.A, ConsoleColor.Red, $"Unable to read from {portName} ({baudRate}): " + ex.Message);
```

Comment:
we should probably skip this logging, because at this step we are tyring to detect a third party device and the result was not detected.

it will get logged anyway when we try to read from it as a normal board

Reply:
Alright, will just make it break resulting in return default.